src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnection.cs

```diff
@@ -1045,95 +1075,202 @@ private static void ParseStatusLine(ReadOnlySpan<byte> line, HttpResponseMessage
+            }
+
+            _readOffset += bytesConsumed;
+            _allowedReadLineBytes -= bytesConsumed;
+            Debug.Assert(_allowedReadLineBytes >= 0);
+
+            return finished;
+        }
+
+        private (bool finished, int bytesConsumed) ParseHeadersCore(Span<byte> buffer, HttpResponseMessage? response, bool isFromTrailer)
+        {
+            int originalBufferLength = buffer.Length;
+
+            while (true)
+            {
+                int colIdx = buffer.IndexOfAny((byte)':', (byte)'\n');
+                if (colIdx < 0) goto needMore;
+
+                if (buffer[colIdx] == '\n')
                 {
-                    // Invalid header line that doesn't contain ':'.
-                    throw new HttpRequestException(SR.Format(SR.net_http_invalid_response_header_line, Encoding.ASCII.GetString(line)));
+                    if ((colIdx == 1 && buffer[0] == '\r') || colIdx == 0)
+                    {
+                        return (finished: true, bytesConsumed: originalBufferLength - buffer.Length + colIdx + 1);
+                    }
+
+                    ThrowForInvalidHeaderLine(buffer);
+                }
+
+                int valueStartIdx = colIdx + 1;
+
+                if ((uint)valueStartIdx >= (uint)buffer.Length) goto needMore;
+
+                Span<byte> valueIter = buffer.Slice(valueStartIdx);
+
+                while (true)
+                {
+                    int lfIdx = valueIter.IndexOf((byte)'\n');
+                    if ((uint)lfIdx >= (uint)valueIter.Length) goto needMore;
+
+                    int crIdx = lfIdx - 1;
+                    int crOrLfIdx = (uint)crIdx < (uint)valueIter.Length && valueIter[crIdx] == '\r'
+                        ? crIdx
+                        : lfIdx;
+
+                    int spIdx = lfIdx + 1;
+                    if ((uint)spIdx >= (uint)valueIter.Length) goto needMore;
+
+                    if (valueIter[spIdx] is not (byte)'\t' and not (byte)' ')
+                    {
+                        // Found the end of the header value.
+
+                        if (response is not null)
+                        {
+                            ReadOnlySpan<byte> headerName = buffer.Slice(0, valueStartIdx - 1);
+                            ReadOnlySpan<byte> headerValue = buffer.Slice(valueStartIdx, buffer.Length - valueIter.Length + crOrLfIdx - valueStartIdx);
+                            AddResponseHeader(headerName, headerValue, response, isFromTrailer);
+                        }
+
+                        buffer = buffer.Slice(buffer.Length - valueIter.Length + spIdx);
+                        break;
+                    }
+
+                    // Found an obs-fold (CRLFHT/CRLFSP).
+                    // Replace the CRLF with SPSP and keep looking for the final newline.
+                    valueIter[crOrLfIdx] = (byte)' ';
+                    valueIter[lfIdx] = (byte)' ';
+
+                    valueIter = valueIter.Slice(spIdx + 1);
```

Comment:
Oh, I see, you're actually slicing/updating valueIter in the case of a folded header; that's why you're doing the more complicated math I commented on earlier.

Reply:
Right, valueIter(ator) is used to handle line folds.
If we decide for .NET 42 that the planet has finally moved away from line folds, this logic could be simplified quite a bit.

Re: your comments above, do you think the logic could be simplified further given that we're supporting line folds?